### PR TITLE
Return correct DeviceInfo from GetDirFsDevice on / path for Btrfs - Fix kubernetes issue #94335

### PR DESCRIPTION
**Pull request by gjkim42**
Fix kubernetes issue #94335

- Issue: kubernetes/kubernetes#94335
- Fix `GetDirFsDevice` to return correct `*DeviceInfo` when "/" is Btrfs
    - Refactor to call new function `mountInfoFromDir` that returns
      `*MountInfo` which `dir` belongs
- Add TestMountInfoFromDir to verify the fix

Signed-off-by: Geonju Kim <rjswn042@gmail.com>

### CODE REVIEW

**Comment by gjkim42** [comment]
@googlebot I signed it!

**Comment by gjkim42** [comment]
add `(*RealFsInfo) mountInfoFromDir`
add a test code

**Comment by gjkim42** [comment]
@iwankgb Would you take a look at this? 
And give me a review or recommend a next step to merge this PR? Thanks!

**Comment by iwankgb** [comment]
@gjkim42 I **think** that as the bug you point at is not part of coming 1.20 release (at least judging by state of the issue) then it will not be merged before Kubernetes release. @bobbypage, can you confirm/deny?
If you want this fix to make into cAdvisor (and Kubernetes) then I would recommend joining [#sig-node](https://kubernetes.slack.com/archives/C0BP8PW9G) and mentioning it there. 
Regarding the review: I am not able to test this PR on real hardware so I'm not particularly keen to review it. Besides this - my powers here are rather limited and I'm not project maintainer.

**Comment by gjkim42** [comment]
@iwankgb Sorry for misunderstanding your role.
I've joined #sig-node and mentioned this issue as you recommended. Thanks for your comment!

@bobbypage Would you take a look at this, please?

**Comment by bobbypage** [comment]
Thanks for the fix, it looks reasonable to me. 

It looks like the main issue here is that the current code doesn't update the mount when it hits pathdir == "/" and that's the main fix here: https://github.com/google/cadvisor/pull/2752/files#diff-c7b54bb28e1654d0ca47364a2c97456c124ee0d3a65854db18d116c509d89136R539. 

The other changes are refactoring of GetDirFsDevice to use a new function `mountInfoFromDir`. Am I understanding the fix correctly? If so, can you please update the PR description to make that a bit more clear on what exact changes / fixes this PR is making?

**Comment by bobbypage** [comment]
Regarding https://github.com/google/cadvisor/pull/2752#issuecomment-744358415

> @gjkim42 I **think** that as the bug you point at is not part of coming 1.20 release (at least judging by state of the issue) then it will not be merged before Kubernetes release. @bobbypage, can you confirm/deny?
> If you want this fix to make into cAdvisor (and Kubernetes) then I would recommend joining [#sig-node](https://kubernetes.slack.com/archives/C0BP8PW9G) and mentioning it there.
> Regarding the review: I am not able to test this PR on real hardware so I'm not particularly keen to review it. Besides this - my powers here are rather limited and I'm not project maintainer.

Yes, after we merge this fix, to actually get the fixed included in k8s will require cutting a new cAdvisor release and vendoring it back to k/k. 

We cut a new cAdvisor release for each k8s release, so the fix will be included in cAdvisor v0.39 which we'll cut for 1.21. If you think this fix is more critical and affecting more users (not sure how widespread BTRFS usage is) and think the fix can be justified to cherrypick into 1.20, I can help cut a patch release of existing cAdvisor so that it can be vendored back into k/k.

**Comment by gjkim42** [comment]
@bobbypage Thanks for your comments!

I updated the PR description as you recommended. Thanks!

---
Regarding the comment,
> We cut a new cAdvisor release for each k8s release, so the fix will be included in cAdvisor v0.39 which we'll cut for 1.21. If you think this fix is more critical and affecting more users (not sure how widespread BTRFS usage is) and think the fix can be justified to cherrypick into 1.20, I can help cut a patch release of existing cAdvisor so that it can be vendored back into k/k.

Although anyone using BTRFS as the root filesystem has encountered this issue, I am not sure this fix is critical enough to be cherry-picked. (In my case, Fedora 33 installer recommended Btrfs as the default.)
I think vendoring it into k8s 1.21 would be OK.

**Comment by bobbypage** [comment]
Thanks for fix, and sounds good re. including this for 1.21.

Changes are LGTM. Just to confirm, were you able to test this change and ensure this fixes the issue on a machine with Btrfs filesystem? I haven't used Btrfs myself :)

**Comment by gjkim42** [comment]
Yes I did!
I have an environment that use Btrfs as a root filesystem.
In my case, I verified that the issue was resolved on my machine.

**Comment by bobbypage** [comment]
Awesome, thanks!

**Review by bobbypage — APPROVED** [review]

**Comment by bobbypage** [comment]
I got a chance to validate the fix myself.

I was able to reproduce https://github.com/kubernetes/kubernetes/issues/94335 on a Fedora 33 VM which uses BTRFs by default.

```
[david@localhost testing]$ uname -a
Linux localhost.localdomain 5.8.15-301.fc33.x86_64 #1 SMP Thu Oct 15 16:58:06 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
[david@localhost testing]$ cat /etc/os-release
NAME=Fedora
VERSION="33 (Workstation Edition)"
ID=fedora
VERSION_ID=33
VERSION_CODENAME=""
PLATFORM_ID="platform:f33"
PRETTY_NAME="Fedora 33 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:33"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f33/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=33
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=33
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
```

BTRFs mounted on `/`

```
[david@localhost testing]$ df -h -T
Filesystem     Type      Size  Used Avail Use% Mounted on
devtmpfs       devtmpfs  970M     0  970M   0% /dev
tmpfs          tmpfs     990M     0  990M   0% /dev/shm
tmpfs          tmpfs     396M  1.4M  395M   1% /run
/dev/sda2      btrfs      31G  6.6G   25G  22% /
/dev/sda2      btrfs      31G  6.6G   25G  22% /home
tmpfs          tmpfs     990M  148K  990M   1% /tmp
/dev/sda1      ext4      976M  183M  727M  21% /boot
tmpfs          tmpfs     198M  148K  198M   1% /run/user/1000
```

Trying out kubelet in standalone mode on 1.20.1:

```
$ mkdir manifests
$ cat > manifests/nginx.yaml << EOF
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx
    imagePullPolicy: Always
EOF

curl -O https://storage.googleapis.com/kubernetes-release/release/v1.20.1/bin/linux/amd64/kubelet
chmod +x kubelet
sudo ./kubelet --cgroup-driver=systemd --pod-manifest-path=$PWD/manifests --fail-swap-on=false
```

and kubelet fails startup with:
```
F0108 14:32:27.432289    4501 kubelet.go:1347] Failed to start ContainerManager failed to get rootfs info: failed to get device for dir "/var/lib/kubelet": could not find device with major: 0, minor: 33 in cached partitions map
goroutine 238 [running]:
```
full stacktrace: https://termbin.com/88qb

I vendored cAdvisor with this fix back into k8s and built kubelet and was able to validate that this PR fixes the issue.

**Pull request merged**
